sql/handler.cc

```diff
@@ -2136,6 +2138,14 @@ int ha_rollback_to_savepoint(THD *thd, SAVEPOINT *sv)
     status_var_increment(thd->status_var.ha_savepoint_rollback_count);
     trans->no_2pc|= ht->prepare == 0;
   }
+
+  /*
+    Write tables map once more for trigger as it will be wiped out in
+    Write_rows_log_event::write_row() on slave.
+  */
+  if (thd->in_sub_stmt == SUB_STMT_TRIGGER)
+    error= write_locked_table_maps(thd, true);
```

Comment:
>One minor issue re. error flag, but I'm not sure whether it's worth fixing. ha_rollback_to_savepoint sets it disregarding any previous value in the flag for each ht->*_rollback call. 

Yes, the logic is if only one storage engine returns error then the whole rollback procedure returns error. As error de-facto is boolean here then "error=" and "error|=" do the same thing.

> The failure of these calls might as well be an assert failure only, as the comments "cannot happen" say, because the main way they would fail in InnoDB is if the savepoint name is not found, which is validated earlier

I would not believe this comment. This is common code which must work for any engine and some abstract engine in vacuum can have any reason to fail rollback procedure, and this code must take it into account.

About innodb. What about this?
trx_rollback_to_savepoint_for_mysql() (): 
	switch (trx->state) {
	case TRX_STATE_NOT_STARTED:
		ut_print_timestamp(stderr);
		fputs("  InnoDB: Error: transaction has a savepoint ", stderr);
		ut_print_name(stderr, trx, FALSE, savep->name);
		fputs(" though it is not started\n", stderr);
		return(DB_ERROR);

I understand your doubts. I ran down to the call stack:
ha_rollback_to_savepoint()
trans_rollback_to_savepoint()
mysql_execute_command()

And error in ha_rollback_to_savepoint() is processed correctly.